keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -646,13 +667,67 @@ Pod Status in response to user changing the desired resources in Pod Spec.
 * At this time, Vertical Pod Autoscaler should not be used with Horizontal Pod
   Autoscaler on CPU, memory. This enhancement does not change that limitation.
 
+### Lifecycle Nuances
+
+* Terminated containers can be "resized" in that the resize is permitted by the API, and the Kubelet
+  will accept the changes. This makes race conditions where the container terminates around the
+  resize "fail open", and prevents a resize of a terminated container from blocking the resize of a
+  running container (see [Atomic Resizes](#atomic-resizes)).
+* Resizing pods in a graceful shutdown state is permitted, and will be actuated best-effort.
+
+### Atomic Resizes
+
+A single resize request can change multiple values, including any or all of:
+* Multiple resource types
+* Requests & Limits
+* Multiple containers
+
+These resource requests & limits can have interdependencies that Kubernetes may not be aware of. For
+example, two containers coordinating work may need to be scaled in tandem. It probably doesn't makes
```

Comment:
Is there a clear, documented use case for "gang resize" or is this speculation?

I know I leveraged the notion of atomic `NodeResourcesFit` predicate when computing resize fit to make the case for pod-level ResizeStatus vs. container-level ResizeStatus with @thockin . But we may be making things worse by building it into the design. 

Consider this:
1. We have a node (4 CPUs, 4Gi memory) with two pods: P1(req: {cpu: 2000m, mem: 3000Mi}) and P2(req: {cpu: 1500m, mem: 1000Mi}).
1. VPA triggers in-place pod resize for P1 updating it to P1(req: {cpu: 3000m, mem: 1000Mi}). But due to atomic resize notion, this request will remain in `Deffered` state keeping P1 at its original resource allocation levels for all types of resources.
1. Some time later, a new pod P3(req: {mem: 2000Mi}) comes along. This pod P3 could have run on the node if the above resize request was deconstructed to P1(req: {cpu: 2000m, mem: 1000Mi}) + P1(req: {cpu: 3000m, mem: 1000Mi})). P1 is still `Deferred` but is now closer to its desired state.

In the above scenario, P3 either remains pending or CA fires up another node just to run P3 - either case increases cost to the customer due to resource waste via over-provisioning.

Reply:
In this case, I am also talking about resizing multiple containers in a single pod with something like the tandem resize constraint (to clarify my “gang resize” terminology above - and yes let’s absolutely leave resizing multiple pods all-or-none ("syndicate resize"? 😓  ) out of scope .. **non-goal**). Also this is not some corner case thing, it is common for app-owners to over-provision their requests e.g ask for way too much memory when app is CPU-intensive. (We have an infra-team resource partly tasked with tuning these and negotiating with app-teams to lower our DC hardware costs and power footprint). Another (contrived) example is a pod’s memory request & limit increase resize-request being withheld despite plenty of memory being available just because CPU increase part of the same resize request is Deferred (leading to OOM-kills)

If those resizes were to come in as multiple discrete requests and we can handle them with the same outcome, then we should be able to handle it coming in as a single request. A complex resize request is eventually broken down to simple [UpdateContainerResources CRI calls](https://github.com/kubernetes/kubernetes/blob/0b91def6bcbdde82206ec779791762fb6bc337af/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L773-L820) to actuate them. I think the same can be done at admit and get the pod as close to desired state as possible. I feel atomic-resize notion at admit can become a footgun that leaves clusters over-provisioned… kinda defeats the purpose of this project.

Thinking out loud, what if we deconstructed a complex resize and admitted what we can (decrease requests will always be admitted), and set status to Deferred or Infeasible if any request cannot be admitted? I see now why @thockin's per-container ResizeStatus makes sense (atleast internally, while keeping pod-level ResizeStatus for VPA to consume). Pod-level resize status will be Infeasible if one or more container requests are Infeasible, and Deferred if one or more requests are Deferred but all are feasible.

I need to think through this very carefully but it’s been a long day and I’m beat rn… however modeling it this way feels like it will work. WDYT?

Reply:
I'm skeptical that trying to do partial resize is useful or correct, and it seems more likely to fail the "least surprise" test.  Kube is level-actuated, kubelet should be driving at the most recent full update, I think.

Reply:
IIUC, level-actuated principle would apply if two resizes R1 (cpu: 1->2) and R2(cpu: 2->3) both arrive before Kubelet has the chance to admit R1. In that case, we do R(cpu: 1->3), if it can be admitted. And all-or-none at pod atomicity makes sense when scheduling and admitting new pod at Kubelet.

In this case, we have R1(mem: 1->2) which is admissible and R2(cpu: 1->2) which is Deferred. If we get [R1,R2], we would admit R1 and defer R2. But if it came in as [R1+R2] or [R2,R1] then we would head-of-line block R1. The scheduler anyways earmarks resources for both R1 & R2 via max(Requests, AllocatedResources). So why withhold R1 by introducing artificial barriers that are not really necessary and waste those resources?

By admitting what we can when those resources are already reserved, we would bring the pod's actual state closer to its desired state than with pod-level atomic resize. We wouldn't hold up a container image update because a Deferred resize prevents a full update, would we?

Reply:
We can't know what was intended by two different resizes, so rather than try to do half of it, I think it's safer and simpler to block the whole transformation.  If we can learn more about why this is important to support, I'd be happy to revisit.

Reply:
+1 on @thockin's points above. I go with the simplicity for now. Getting the in-place pod resizing feature to beta is a significant milestone to us. Focusing on the core functionality and addressing the most common use cases with the atomic approach first allows for a faster beta release. Once the feature is released as beta, we can collect the feedback from users. We can evolve the feature with the partial resizes based on the feedbacks. @vinaykul WDYT?

Reply:
I'm not in favor of this, defeats the purpose of the project. Make it a GA blocker?

Reply:
> defeats the purpose of the project

That's a huge statement.  What we have here is a relatively minor case (and if not, something unforeseen is happening).  IMO, having a clearly defined and safe semantic for such a corner case is more important than being exactly right in every case.

Reply:
Thanks Vinay! 

@tallclair 
1) Let's document clearly the limitations of the atomic approach and indicate the partial resizes are not yet supported and the potential for resource waste in certain scenarios. The plan is calling out the inputs from the users  who are using the coming beta feature. 
2) Let's also include re-evaluating this for GA too. Based on the data and feedback we have, we can re-evaluate the need for partial resizes before GA. If the data shows significant resource waste or user frustration, prioritize implementing partial resizes as a pre-requisite for GA.

Reply:
+1 to Dawn's suggestions. I'll update the KEP to include this.

I think we also need to reevaluate this in the context of pod-level resources, so I'd rather wait on a solution.